src/crystal/system/win32/iocp.cr

```diff
@@ -136,49 +122,46 @@ module Crystal::IOCP
     end
 
     protected def schedule(&)
-      case @state
-      when .started?
-        yield @fiber
-        done!
-      when .cancelled?
-        @@canceled.delete(self)
-      else
-        raise Exception.new("Invalid state #{@state}")
-      end
-    end
-
-    protected def done
-      case @state
-      when .started?
-        # https://learn.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-cancelioex
-        # > The application must not free or reuse the OVERLAPPED structure
-        # associated with the canceled I/O operations until they have completed
-        if LibC.CancelIoEx(@handle, self) != 0
-          @state = :cancelled
-          @@canceled.push(self) # to increase lifetime
-        end
-      end
+      done!
+      yield @fiber
     end
 
     def done!
+      @fiber.cancel_timeout
       @state = :done
     end
 
+    def cancel! : Bool
+      # Microsoft documentation:
+      # The application must not free or reuse the OVERLAPPED structure
+      # associated with the canceled I/O operations until they have completed
+      # (this does not apply to asynchronous operations that finished
+      # synchronously, as nothing would be queued to the IOCP)
+      ret = LibC.CancelIoEx(@handle, self)
+      if ret.zero?
+        case error = WinError.value
+        when .error_not_found?
+          # Operation has already completed, do nothing
+          return false
+        else
+          raise RuntimeError.from_os_error("CancelIOEx", os_error: error)
+        end
+      end
+      true
+    end
+
     def wait_for_completion(timeout)
       if timeout
-        timeout_event = Crystal::IOCP::Event.new(Fiber.current)
-        timeout_event.add(timeout)
+        sleep timeout
       else
-        timeout_event = Crystal::IOCP::Event.new(Fiber.current, Time::Span::MAX)
+        Fiber.suspend
       end
-      # memoize event loop to make sure that we still target the same instance
-      # after wakeup (guaranteed by current MT model but let's be future proof)
-      event_loop = Crystal::EventLoop.current
-      event_loop.enqueue(timeout_event)
-
-      Fiber.suspend
 
-      event_loop.dequeue(timeout_event)
+      unless @state.done?
+        if cancel!
+          Fiber.suspend
```

Comment:
**suggestion**: maybe explain why we need to suspend the fiber?

Reply:
**suggestion**: we only need to cancel the operation when the fiber awoke from a timeout, so inside the `if timeout` branch above. The other branch will only ever resume the fiber when the operation was completed.

Or is that to protect against spurious wakeups?

Reply:
Yeah, the cancellation is only relevant when there is a timeout.
Dunno if it's worth keeping it for both branches as a sanity measure 🤔 I don't see any immediate reason to expect a spurious wakeup.

Reply:
I guess there could be use cases for external cancellation, though. We don't need to explicitly cancel pending operations upon closing a handle because IOCP does that automatically. But maybe there could be other situations where we want to cancel a pending operation from a different fiber? 🤔